README.md
### GADGET-2 code with Oscillating ADM model (for ligth scalar and vector mediators) 

Code used in:

Cline, J. M., Gambini, G., McDermott S. D., Puel, M., *Late-Time Dark Matter Oscillations and the Core-Cusp Problem*, J. High Energ. Phys. 2021, 223 (2021), https://doi.org/10.1007/JHEP04(2021)223

and based on GADGET-2, written by Volker Springel (https://wwwmpa.mpa-garching.mpg.de/gadget/). Please cite the initial paper, bseides the reference above, if using this code.

Author of modifications: Matteo Puel @2020


## Compilation 
The compilation file is **Makefile**, which is the same as in GADGET-2 (see corresponding manual). New parameters have been added and here below we provide a description.

ADM_MODEL_ON : Master flag to allow to turn on our model properties.
- VECTOR_MEDIATOR : Massive light vector mediator interaction. It includes the Sommerfeld enhancement factor for s-wave annihilation
- SCALAR_MEDIATOR : Massive light scalar mediator interaction. It includes the Sommerfeld enhancement factor for s-wave annihilation
- OSCILLATION_DM : Activate oscillations
- ANNIHILATION_DM : Activate annihilations (just s-wave cross section). It requires oscillation probability for Dirac DM. 
- SCATTERING_DM : Activate scatterings. Standard is the viscosity cross section with isotropic scattering
-TRANSFER_CROSS_SECTION : Requires SCATTERING_DM to be active! Replace the viscosity cross section with the original transfer cross section. Still isotropic scattering
- MODIFIED_TRANSFER_CROSS_SECTION : Requires SCATTERING_DM to be active! Replace the viscosity cross section with the modified transfer cross section. Still isotropic scattering
- VELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION : Requires SCATTERING_DM to be active! Polar angle and cross section values stored in tables given in input. Anisotropic scattering
- LATE_TIME_PHASE_CHANGE : Requires SCATTERING_DM to be active! Add random phase of scattered particles if cosmic time > 1/MajoranaMass
- OUTPUTDT_SCATTER : Requires SCATTERING_DM to be active! Include the current dt_scatter variable of all particles in the snapshot files (similar to OUTPUTTIMESTEP). dt_scatter is also read from the snapshot if simulation has RestartFlag == 2
- TEST_UNIFORM_BKG : Test 1: scattering of cube to uniform background (no gravity, just one scattering per each particle)
- TEST_HERNQUIST_HALO_SCATT : Test 2: scattering in an isolated Hernquist halo (no change in the particle velocity)
- TEST_HERNQUIST_HALO_ANN : Test 2: scattering in an isolated Hernquist halo (no remove annihilated particles)


Additional paramterfile options:
- MediatorDMmassratio : Mass ratio between mediator and DM, equivalent to dimensionless w (only for VECTOR_MEDIATOR)
- VectorOrScalarCoupling : Dimensionless coupling between vector or scalar mediator and DM, equivalent to \alpha_D
- MajoranaMass : Majorana mass delta_m in eV causing DM oscillations (only for OSCILLATION_DM)
- AnnihilationCrossSectionSwave : The DM velocity-averaged annihilation cross section (s-wave, i.e constant, velocity-independent) per unit DM mass in cm^3/g/sec
- AnnihilateSearchRadius : The neighbour search radius for DM annihilation (h_A) in comoving code length units
- AnnihilateSearchRadiusMaxPhys : The maximum physical search radius (a \times h_A) for DM annihilation in code length units
- ScatteringCrossSection : The DM scattering cross section per unit mass in cm^2/g
- ScatterSearchRadius : The neighbour search radius for DM scattering (h_S) in comoving code length units
- ScatterSearchRadiusMaxPhys : The maximum physical search radius (a \times h_S) for DM scattering in code length units
- ThetaTableFilename : Name of the file containing polar angles drawn from an anisotropic velocity-dependent differential cross section (only for VELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION)
- CrossSectionTableFilename : Name of the file containing total cross sections drawn from an anisotropic velocity-dependent differential cross section as a function of velocity (only for VELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION)
- Nv : The number of velocities at which the cross-section and a sample of thetas are defined (only for VELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION)
- Ntheta : The number of samples drawn from the differential cross-section at each velocity (only for VELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION)
- vmin : The minumum velocity (in km/s) at which the cross-section is defined. For velocities v<vmin, the cross-section is set equal to its value at vmin (only for VELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION)
- vmax : The maximum velocity (in km/s) at which the cross-section is defined. For velocities v>vmax, the cross-section is set to 0 (only for VELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION)
- ProbabilityTol : Maximum probability of scattering, used to introduce a DM scattering limiter (not for annihilation because the two annihilating particles disappear)

## Bibtex
@article{Cline:2020gon,
    author = "Cline, James M. and Gambini, Guillermo and Mcdermott, Samuel D. and Puel, Matteo",
    title = "{Late-Time Dark Matter Oscillations and the Core-Cusp Problem}",
    eprint = "2010.12583",
    archivePrefix = "arXiv",
    primaryClass = "hep-ph",
    reportNumber = "FNAL-PUB-20-556-T, FERMILAB-PUB-20-556-T",
    doi = "10.1007/JHEP04(2021)223",
    journal = "JHEP",
    volume = "04",
    pages = "223",
    year = "2021"
}



CHANGELOG
### Solve some warnings ###
"allocate.c": line 119, substitute "printf("\nAllocated %g MByte for particle storage. %lu\n\n", bytes_tot / (1024.0 * 1024.0), sizeof(struct particle_data));"
"allocate.c": line 129, substitute "printf("failed to allocate memory for `SphP' (%g MB) %lu.\n", bytes / (1024.0 * 1024.0), sizeof(struct sph_particle_data));"
"allocate.c": line 136, substitute "printf("Allocated %g MByte for storage of SPH data. %lu\n\n", bytes_tot / (1024.0 * 1024.0), sizeof(struct sph_particle_data));"

"forcetree.c": line 2668, substitute printf("\nAllocated %g MByte for BH-tree. %lu\n\n", allbytes / (1024.0 * 1024.0), sizeof(struct NODE) + sizeof(struct extNODE));"
"gravtree.c": line 32, substitute "int *numnodeslist, maxnumnodes, nexport, *numlist, *nrecv;"
"gravtree.c": line 40, substitute "int *noffset, *nbuffer, *nsend, *nsend_local, *ndonelist;"


### Changes to include DM Oscillations, Annihilations and Scatterings ###

write README.md and README.txt from scratch with description of implementations of our model in the code

"Makefile": line 51, add "OPT   +=  -DADM_MODEL_ON # scalar mediator interaction as the default model"
"Makefile": line 52, add "OPT   +=  -DVECTOR_MEDIATOR"
"Makefile": line 53, add "OPT   +=  -DSCALAR_MEDIATOR"
"Makefile": line 54, add "OPT   +=  -DOSCILLATION_DM"
"Makefile": line 55, add "OPT   +=  -DANNIHILATION_DM"
"Makefile": line 56, add "OPT   +=  -DSCATTERING_DM"
"Makefile": line 57, add "OPT   +=  -DTRANSFER_CROSS_SECTION"
"Makefile": line 58, add "OPT   +=  -DMODIFIED_TRANSFER_CROSS_SECTION"
"Makefile": line 59, add "OPT   +=  -DVELOCITY_ANGULAR_DEPENDENT_CROSS_SECTION"
"Makefile": line 60, add "OPT   +=  -DLATE_TIME_PHASE_CHANGE"
"Makefile": line 61, add "OPT   +=  -DOUTPUTDT_SCATTER"
"Makefile": line 62, add "OPT   +=  -DTEST_UNIFORM_BKG # test 1"
"Makefile": line 63, add "OPT   +=  -DTEST_HERNQUIST_HALO_SCATT # test 2 (scattering)"
"Makefile": line 64, add "OPT   +=  -DTEST_HERNQUIST_HALO_ANN # test 2 (annihilation)"
"Makefile": line 273, add "ifeq (ADM_MODEL_ON,$(findstring ADM_MODEL_ON, $(OPT)))"
"Makefile": line 274, add "OBJS += adm_model.o"
"Makefile": line 275, add "endif"

"ADMmodel.param": line 133, add "% Vector or Scalar mediator model"
"ADMmodel.param": line 134, add "MediatorDMmassratio		0.3		% w = m_V / m_DM"
"ADMmodel.param": line 135, add "VectorOrScalarCoupling			0.02	% alpha_D"
"ADMmodel.param": line 137, add "% Oscillation"
"ADMmodel.param": line 138, add "MajoranaMass		1.0e-31		% delta_m in eV/c^2"
"ADMmodel.param": line 140, add "% Annihilation"
"ADMmodel.param": line 141, add "AnnihilationCrossSectionSwave			1.00		% cm^3/g/sec"
"ADMmodel.param": line 142, add "AnnihilateSearchRadius  			0.0003 		% 0.01562496"
"ADMmodel.param": line 143, add "AnnihilateSearchRadiusMaxPhys 		0.00008 	% 0.004"
"ADMmodel.param": line 145, add "% Scattering"
"ADMmodel.param": line 146, add "ScatteringCrossSection			1.00		% cm^2/g"
"ADMmodel.param": line 147, add "ScatterSearchRadius  			0.0003 		% 0.01562496"
"ADMmodel.param": line 148, add "ScatterSearchRadiusMaxPhys 		0.00008 	% 0.004"
"ADMmodel.param": line 150, add "ThetaTableFilename  /.../theta_tables_2D/yukawa_w1_th.txt"
"ADMmodel.param": line 151, add "CrossSectionTableFilename /.../theta_tables_2D/yukawa_w1_cs.txt"
"ADMmodel.param": line 152, add "Nv 		1000"
"ADMmodel.param": line 153, add "Ntheta 		1000"
"ADMmodel.param": line 154, add "vmin 	0.560 % to get w=560 km/s, with w=1 in CrossSection and Theta files"
"ADMmodel.param": line 155, add "vmax 	560000"
"ADMmodel.param": line 157, add "ProbabilityTol    0.2 	% 0.2 (Rocha), to avoid multiple scattering in each timestep (in principle it should be at most 1.0)"

"peano.c": from line 211 to line 260, add "const unsigned char rottable3[48][8] = {
{36, 28, 25, 27, 10, 10, 25, 27},
{29, 11, 24, 24, 37, 11, 26, 26},
{8, 8, 25, 27, 30, 38, 25, 27},
{9, 39, 24, 24, 9, 31, 26, 26},
{40, 24, 44, 32, 40, 6, 44, 6},
{25, 7, 33, 7, 41, 41, 45, 45},
{4, 42, 4, 46, 26, 42, 34, 46},
{43, 43, 47, 47, 5, 27, 5, 35},
{33, 35, 36, 28, 33, 35, 2, 2},
{32, 32, 29, 3, 34, 34, 37, 3},
{33, 35, 0, 0, 33, 35, 30, 38},
{32, 32, 1, 39, 34, 34, 1, 31},
{24, 42, 32, 46, 14, 42, 14, 46},
{43, 43, 47, 47, 25, 15, 33, 15},
{40, 12, 44, 12, 40, 26, 44, 34},
{13, 27, 13, 35, 41, 41, 45, 45},
{28, 41, 28, 22, 38, 43, 38, 22},
{42, 40, 23, 23, 29, 39, 29, 39},
{41, 36, 20, 36, 43, 30, 20, 30},
{37, 31, 37, 31, 42, 40, 21, 21},
{28, 18, 28, 45, 38, 18, 38, 47},
{19, 19, 46, 44, 29, 39, 29, 39},
{16, 36, 45, 36, 16, 30, 47, 30},
{37, 31, 37, 31, 17, 17, 46, 44},
{12, 4, 1, 3, 34, 34, 1, 3},
{5, 35, 0, 0, 13, 35, 2, 2},
{32, 32, 1, 3, 6, 14, 1, 3},
{33, 15, 0, 0, 33, 7, 2, 2},
{16, 0, 20, 8, 16, 30, 20, 30},
{1, 31, 9, 31, 17, 17, 21, 21},
{28, 18, 28, 22, 2, 18, 10, 22},
{19, 19, 23, 23, 29, 3, 29, 11},
{9, 11, 12, 4, 9, 11, 26, 26},
{8, 8, 5, 27, 10, 10, 13, 27},
{9, 11, 24, 24, 9, 11, 6, 14},
{8, 8, 25, 15, 10, 10, 25, 7},
{0, 18, 8, 22, 38, 18, 38, 22},
{19, 19, 23, 23, 1, 39, 9, 39},
{16, 36, 20, 36, 16, 2, 20, 10},
{37, 3, 37, 11, 17, 17, 21, 21},
{4, 17, 4, 46, 14, 19, 14, 46},
{18, 16, 47, 47, 5, 15, 5, 15},
{17, 12, 44, 12, 19, 6, 44, 6},
{13, 7, 13, 7, 18, 16, 45, 45},
{4, 42, 4, 21, 14, 42, 14, 23},
{43, 43, 22, 20, 5, 15, 5, 15},
{40, 12, 21, 12, 40, 6, 23, 6},
{13, 7, 13, 7, 41, 41, 22, 20}
};"
"peano.c": from line 262 to line 311, add "const unsigned char subpix3[48][8] = {
{0, 7, 1, 6, 3, 4, 2, 5},
{7, 4, 6, 5, 0, 3, 1, 2},
{4, 3, 5, 2, 7, 0, 6, 1},
{3, 0, 2, 1, 4, 7, 5, 6},
{1, 0, 6, 7, 2, 3, 5, 4},
{0, 3, 7, 4, 1, 2, 6, 5},
{3, 2, 4, 5, 0, 1, 7, 6},
{2, 1, 5, 6, 3, 0, 4, 7},
{6, 1, 7, 0, 5, 2, 4, 3},
{1, 2, 0, 3, 6, 5, 7, 4},
{2, 5, 3, 4, 1, 6, 0, 7},
{5, 6, 4, 7, 2, 1, 3, 0},
{7, 6, 0, 1, 4, 5, 3, 2},
{6, 5, 1, 2, 7, 4, 0, 3},
{5, 4, 2, 3, 6, 7, 1, 0},
{4, 7, 3, 0, 5, 6, 2, 1},
{6, 7, 5, 4, 1, 0, 2, 3},
{7, 0, 4, 3, 6, 1, 5, 2},
{0, 1, 3, 2, 7, 6, 4, 5},
{1, 6, 2, 5, 0, 7, 3, 4},
{2, 3, 1, 0, 5, 4, 6, 7},
{3, 4, 0, 7, 2, 5, 1, 6},

peano.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <mpi.h>

#include "allvars.h"
#include "proto.h"

/*! \file peano.c 
 *  \brief Routines to compute a Peano-Hilbert order
 *
 *  This file contains routines to compute Peano-Hilbert keys, and to put the
 *  particle data into the order of these keys, i.e. into the order of a
 *  space-filling fractal curve.
 */


static struct peano_hilbert_data
{
  peanokey key;
  int index;
}
 *mp;

static int *Id;


/*! This function puts the particles into Peano-Hilbert order by sorting them
 *  according to their keys. The latter half already been computed in the
 *  domain decomposition. Since gas particles need to stay at the beginning of
 *  the particle list, they are sorted as a separate block.
 */
void peano_hilbert_order(void)
{
  int i;

  if(ThisTask == 0)
    printf("begin Peano-Hilbert order...\n");

  if(N_gas)
    {
      mp = malloc(sizeof(struct peano_hilbert_data) * N_gas);
      Id = malloc(sizeof(int) * N_gas);

      for(i = 0; i < N_gas; i++)
	{
	  mp[i].index = i;
	  mp[i].key = Key[i];
	}

      qsort(mp, N_gas, sizeof(struct peano_hilbert_data), compare_key);

      for(i = 0; i < N_gas; i++)
	Id[mp[i].index] = i;

      reorder_gas();

      free(Id);
      free(mp);
    }


  if(NumPart - N_gas > 0)
    {
      mp = malloc(sizeof(struct peano_hilbert_data) * (NumPart - N_gas));
      mp -= (N_gas);

      Id = malloc(sizeof(int) * (NumPart - N_gas));
      Id -= (N_gas);

      for(i = N_gas; i < NumPart; i++)
	{
	  mp[i].index = i;
	  mp[i].key = Key[i];
	}

      qsort(mp + N_gas, NumPart - N_gas, sizeof(struct peano_hilbert_data), compare_key);

      for(i = N_gas; i < NumPart; i++)
	Id[mp[i].index] = i;

      reorder_particles();

      Id += N_gas;
      free(Id);
      mp += N_gas;
      free(mp);
    }

  if(ThisTask == 0)
    printf("Peano-Hilbert done.\n");
}


/*! This function is a comparison kernel for sorting the Peano-Hilbert keys.
 */
int compare_key(const void *a, const void *b)
{
  if(((struct peano_hilbert_data *) a)->key < (((struct peano_hilbert_data *) b)->key))
    return -1;

  if(((struct peano_hilbert_data *) a)->key > (((struct peano_hilbert_data *) b)->key))
    return +1;

  return 0;
}


/*! This function brings the gas particles into the same order as the sorted
 *  keys. (The sort is first done only on the keys themselves and done
 *  directly on the gas particles in order to reduce the amount of data that
 *  needs to be moved in memory. Only once the order is established, the gas
 *  particles are rearranged, such that each particle has to be moved at most
 *  once.)
 */
void reorder_gas(void)
{
  int i;
  struct particle_data Psave, Psource;
  struct sph_particle_data SphPsave, SphPsource;
  int idsource, idsave, dest;

  for(i = 0; i < N_gas; i++)
    {
      if(Id[i] != i)
	{
	  Psource = P[i];
	  SphPsource = SphP[i];

	  idsource = Id[i];
	  dest = Id[i];

	  do
	    {
	      Psave = P[dest];
	      SphPsave = SphP[dest];
	      idsave = Id[dest];

	      P[dest] = Psource;
	      SphP[dest] = SphPsource;
	      Id[dest] = idsource;

	      if(dest == i)
		break;

	      Psource = Psave;
	      SphPsource = SphPsave;
	      idsource = idsave;

	      dest = idsource;
	    }
	  while(1);
	}
    }
}


/*! This function brings the collisionless particles into the same order as
 *  the sorted keys. (The sort is first done only on the keys themselves and
 *  done directly on the particles in order to reduce the amount of data that
 *  needs to be moved in memory. Only once the order is established, the
 *  particles are rearranged, such that each particle has to be moved at most
 *  once.)
 */
void reorder_particles(void)
{
  int i;
  struct particle_data Psave, Psource;
  int idsource, idsave, dest;

  for(i = N_gas; i < NumPart; i++)
    {
      if(Id[i] != i)
	{
	  Psource = P[i];
	  idsource = Id[i];

	  dest = Id[i];

	  do
	    {
	      Psave = P[dest];
	      idsave = Id[dest];

	      P[dest] = Psource;
	      Id[dest] = idsource;

	      if(dest == i)
		break;

	      Psource = Psave;
	      idsource = idsave;

	      dest = idsource;
	    }
	  while(1);
	}
    }
}




// MPUEL: taken from PM-Gadget (https://github.com/sbird/MP-Gadget3) to speed the simulation

/*  The following rewrite of the original function
 *  peano_hilbert_key_old() has been written by MARTIN REINECKE. 
 *  It is about a factor 2.3 - 2.5 faster than Volker's old routine!
 */
const unsigned char rottable3[48][8] = {
  {36, 28, 25, 27, 10, 10, 25, 27},
  {29, 11, 24, 24, 37, 11, 26, 26},
  {8, 8, 25, 27, 30, 38, 25, 27},
  {9, 39, 24, 24, 9, 31, 26, 26},
  {40, 24, 44, 32, 40, 6, 44, 6},
  {25, 7, 33, 7, 41, 41, 45, 45},
  {4, 42, 4, 46, 26, 42, 34, 46},
  {43, 43, 47, 47, 5, 27, 5, 35},
  {33, 35, 36, 28, 33, 35, 2, 2},
  {32, 32, 29, 3, 34, 34, 37, 3},
  {33, 35, 0, 0, 33, 35, 30, 38},
  {32, 32, 1, 39, 34, 34, 1, 31},
  {24, 42, 32, 46, 14, 42, 14, 46},
  {43, 43, 47, 47, 25, 15, 33, 15},
  {40, 12, 44, 12, 40, 26, 44, 34},
  {13, 27, 13, 35, 41, 41, 45, 45},
  {28, 41, 28, 22, 38, 43, 38, 22},
  {42, 40, 23, 23, 29, 39, 29, 39},
  {41, 36, 20, 36, 43, 30, 20, 30},
  {37, 31, 37, 31, 42, 40, 21, 21},
  {28, 18, 28, 45, 38, 18, 38, 47},
  {19, 19, 46, 44, 29, 39, 29, 39},
  {16, 36, 45, 36, 16, 30, 47, 30},
  {37, 31, 37, 31, 17, 17, 46, 44},
  {12, 4, 1, 3, 34, 34, 1, 3},
  {5, 35, 0, 0, 13, 35, 2, 2},
  {32, 32, 1, 3, 6, 14, 1, 3},
  {33, 15, 0, 0, 33, 7, 2, 2},
  {16, 0, 20, 8, 16, 30, 20, 30},
  {1, 31, 9, 31, 17, 17, 21, 21},
  {28, 18, 28, 22, 2, 18, 10, 22},
  {19, 19, 23, 23, 29, 3, 29, 11},
  {9, 11, 12, 4, 9, 11, 26, 26},
  {8, 8, 5, 27, 10, 10, 13, 27},
  {9, 11, 24, 24, 9, 11, 6, 14},
  {8, 8, 25, 15, 10, 10, 25, 7},
  {0, 18, 8, 22, 38, 18, 38, 22},
  {19, 19, 23, 23, 1, 39, 9, 39},
  {16, 36, 20, 36, 16, 2, 20, 10},
  {37, 3, 37, 11, 17, 17, 21, 21},
  {4, 17, 4, 46, 14, 19, 14, 46},
  {18, 16, 47, 47, 5, 15, 5, 15},
  {17, 12, 44, 12, 19, 6, 44, 6},
  {13, 7, 13, 7, 18, 16, 45, 45},
  {4, 42, 4, 21, 14, 42, 14, 23},
  {43, 43, 22, 20, 5, 15, 5, 15},
  {40, 12, 21, 12, 40, 6, 23, 6},
  {13, 7, 13, 7, 41, 41, 22, 20}
};

const unsigned char subpix3[48][8] = {
  {0, 7, 1, 6, 3, 4, 2, 5},
  {7, 4, 6, 5, 0, 3, 1, 2},
  {4, 3, 5, 2, 7, 0, 6, 1},
  {3, 0, 2, 1, 4, 7, 5, 6},
  {1, 0, 6, 7, 2, 3, 5, 4},
  {0, 3, 7, 4, 1, 2, 6, 5},
  {3, 2, 4, 5, 0, 1, 7, 6},
  {2, 1, 5, 6, 3, 0, 4, 7},
  {6, 1, 7, 0, 5, 2, 4, 3},
  {1, 2, 0, 3, 6, 5, 7, 4},
  {2, 5, 3, 4, 1, 6, 0, 7},
  {5, 6, 4, 7, 2, 1, 3, 0},
  {7, 6, 0, 1, 4, 5, 3, 2},
  {6, 5, 1, 2, 7, 4, 0, 3},
  {5, 4, 2, 3, 6, 7, 1, 0},
  {4, 7, 3, 0, 5, 6, 2, 1},
  {6, 7, 5, 4, 1, 0, 2, 3},
  {7, 0, 4, 3, 6, 1, 5, 2},
  {0, 1, 3, 2, 7, 6, 4, 5},
  {1, 6, 2, 5, 0, 7, 3, 4},
  {2, 3, 1, 0, 5, 4, 6, 7},
  {3, 4, 0, 7, 2, 5, 1, 6},
  {4, 5, 7, 6, 3, 2, 0, 1},
  {5, 2, 6, 1, 4, 3, 7, 0},
  {7, 0, 6, 1, 4, 3, 5, 2},
  {0, 3, 1, 2, 7, 4, 6, 5},
  {3, 4, 2, 5, 0, 7, 1, 6},
  {4, 7, 5, 6, 3, 0, 2, 1},
  {6, 7, 1, 0, 5, 4, 2, 3},
  {7, 4, 0, 3, 6, 5, 1, 2},
  {4, 5, 3, 2, 7, 6, 0, 1},
  {5, 6, 2, 1, 4, 7, 3, 0},
  {1, 6, 0, 7, 2, 5, 3, 4},
  {6, 5, 7, 4, 1, 2, 0, 3},
  {5, 2, 4, 3, 6, 1, 7, 0},
  {2, 1, 3, 0, 5, 6, 4, 7},
  {0, 1, 7, 6, 3, 2, 4, 5},
  {1, 2, 6, 5, 0, 3, 7, 4},
  {2, 3, 5, 4, 1, 0, 6, 7},
  {3, 0, 4, 7, 2, 1, 5, 6},
  {1, 0, 2, 3, 6, 7, 5, 4},
  {0, 7, 3, 4, 1, 6, 2, 5},
  {7, 6, 4, 5, 0, 1, 3, 2},
  {6, 1, 5, 2, 7, 0, 4, 3},
  {5, 4, 6, 7, 2, 3, 1, 0},
  {4, 3, 7, 0, 5, 2, 6, 1},
  {3, 2, 0, 1, 4, 5, 7, 6},
  {2, 5, 1, 6, 3, 4, 0, 7}
};

/*! This function computes a Peano-Hilbert key for an integer triplet (x,y,z),
  *  with x,y,z in the range between 0 and 2^bits-1.
  */
peanokey peano_hilbert_key(int x, int y, int z, int bits)
{
  int mask;
  unsigned char rotation = 0;
  peanokey key = 0;

  for(mask = 1 << (bits - 1); mask > 0; mask >>= 1)
    {
      unsigned char pix = ((x & mask) ? 4 : 0) | ((y & mask) ? 2 : 0) | ((z & mask) ? 1 : 0);

      key <<= 3;
      key |= subpix3[rotation][pix];
      rotation = rottable3[rotation][pix];
    }

  return key;
}


/* the following is not needed in the current version of the code */


static int quadrants[24][2][2][2] = {
  /* rotx=0, roty=0-3 */
  {{{0, 7}, {1, 6}}, {{3, 4}, {2, 5}}},
  {{{7, 4}, {6, 5}}, {{0, 3}, {1, 2}}},
  {{{4, 3}, {5, 2}}, {{7, 0}, {6, 1}}},
  {{{3, 0}, {2, 1}}, {{4, 7}, {5, 6}}},
  /* rotx=1, roty=0-3 */
  {{{1, 0}, {6, 7}}, {{2, 3}, {5, 4}}},
  {{{0, 3}, {7, 4}}, {{1, 2}, {6, 5}}},
  {{{3, 2}, {4, 5}}, {{0, 1}, {7, 6}}},
  {{{2, 1}, {5, 6}}, {{3, 0}, {4, 7}}},
  /* rotx=2, roty=0-3 */
  {{{6, 1}, {7, 0}}, {{5, 2}, {4, 3}}},
  {{{1, 2}, {0, 3}}, {{6, 5}, {7, 4}}},
  {{{2, 5}, {3, 4}}, {{1, 6}, {0, 7}}},
  {{{5, 6}, {4, 7}}, {{2, 1}, {3, 0}}},
  /* rotx=3, roty=0-3 */
  {{{7, 6}, {0, 1}}, {{4, 5}, {3, 2}}},
  {{{6, 5}, {1, 2}}, {{7, 4}, {0, 3}}},
  {{{5, 4}, {2, 3}}, {{6, 7}, {1, 0}}},
  {{{4, 7}, {3, 0}}, {{5, 6}, {2, 1}}},
  /* rotx=4, roty=0-3 */
  {{{6, 7}, {5, 4}}, {{1, 0}, {2, 3}}},
  {{{7, 0}, {4, 3}}, {{6, 1}, {5, 2}}},
  {{{0, 1}, {3, 2}}, {{7, 6}, {4, 5}}},
  {{{1, 6}, {2, 5}}, {{0, 7}, {3, 4}}},
  /* rotx=5, roty=0-3 */
  {{{2, 3}, {1, 0}}, {{5, 4}, {6, 7}}},
  {{{3, 4}, {0, 7}}, {{2, 5}, {1, 6}}},
  {{{4, 5}, {7, 6}}, {{3, 2}, {0, 1}}},
  {{{5, 2}, {6, 1}}, {{4, 3}, {7, 0}}}
};


static int rotxmap_table[24] = { 4, 5, 6, 7, 8, 9, 10, 11,
  12, 13, 14, 15, 0, 1, 2, 3, 17, 18, 19, 16, 23, 20, 21, 22
};

static int rotymap_table[24] = { 1, 2, 3, 0, 16, 17, 18, 19,
  11, 8, 9, 10, 22, 23, 20, 21, 14, 15, 12, 13, 4, 5, 6, 7
};

static int rotx_table[8] = { 3, 0, 0, 2, 2, 0, 0, 1 };
static int roty_table[8] = { 0, 1, 1, 2, 2, 3, 3, 0 };

static int sense_table[8] = { -1, -1, -1, +1, +1, -1, -1, -1 };

static int flag_quadrants_inverse = 1;
static char quadrants_inverse_x[24][8];
static char quadrants_inverse_y[24][8];
static char quadrants_inverse_z[24][8];

// MPUEL: replacement peano_hilbert_key() -> peano_hilbert_key_old()

/*! This function computes a Peano-Hilbert key for an integer triplet (x,y,z),
 *  with x,y,z in the range between 0 and 2^bits-1.
 */
peanokey peano_hilbert_key_old(int x, int y, int z, int bits)
{
  int i, quad, bitx, bity, bitz;
  int mask, rotation, rotx, roty, sense;
  peanokey key;


  mask = 1 << (bits - 1);
  key = 0;
  rotation = 0;
  sense = 1;


  for(i = 0; i < bits; i++, mask >>= 1)
    {
      bitx = (x & mask) ? 1 : 0;
      bity = (y & mask) ? 1 : 0;
      bitz = (z & mask) ? 1 : 0;

      quad = quadrants[rotation][bitx][bity][bitz];

      key <<= 3;
      key += (sense == 1) ? (quad) : (7 - quad);

      rotx = rotx_table[quad];
      roty = roty_table[quad];
      sense *= sense_table[quad];

      while(rotx > 0)
	{
	  rotation = rotxmap_table[rotation];
	  rotx--;
	}

      while(roty > 0)
	{
	  rotation = rotymap_table[rotation];
	  roty--;
	}
    }

  return key;
}


/*! This function computes for a given Peano-Hilbert key, the inverse,
 *  i.e. the integer triplet (x,y,z) with a Peano-Hilbert key equal to the
 *  input key. (This functionality is actually not needed in the present
 *  code.)
 */
void peano_hilbert_key_inverse(peanokey key, int bits, int *x, int *y, int *z)
{
  int i, keypart, bitx, bity, bitz, mask, quad, rotation, shift;
  char sense, rotx, roty;

  if(flag_quadrants_inverse)
    {
      flag_quadrants_inverse = 0;
      for(rotation = 0; rotation < 24; rotation++)
        for(bitx = 0; bitx < 2; bitx++)
          for(bity = 0; bity < 2; bity++)
            for(bitz = 0; bitz < 2; bitz++)
              {
                quad = quadrants[rotation][bitx][bity][bitz];
                quadrants_inverse_x[rotation][quad] = bitx;
                quadrants_inverse_y[rotation][quad] = bity;
                quadrants_inverse_z[rotation][quad] = bitz;
              }
    }

  shift = 3 * (bits - 1);
  mask = 7 << shift;

  rotation = 0;
  sense = 1;

  *x = *y = *z = 0;

  for(i = 0; i < bits; i++, mask >>= 3, shift -= 3)
    {
      keypart = (key & mask) >> shift;

      quad = (sense == 1) ? (keypart) : (7 - keypart);

      *x = (*x << 1) + quadrants_inverse_x[rotation][quad];
      *y = (*y << 1) + quadrants_inverse_y[rotation][quad];
      *z = (*z << 1) + quadrants_inverse_z[rotation][quad];

      rotx = rotx_table[quad];
      roty = roty_table[quad];
      sense *= sense_table[quad];

      while(rotx > 0)
        {
          rotation = rotxmap_table[rotation];
          rotx--;
        }

      while(roty > 0)
        {
          rotation = rotymap_table[rotation];
          roty--;
        }
    }
}

